### ISSUE TITLE
Numerical issues with GammaPoisson (i.e, negative binomial) for small rates

### ISSUE THREAD

**Issue by dsheldon** [pyro-ppl/numpyro#563]
I was looking into the use of different observation models for large discrete counts, including the negative binomial distribution, which is equivalent to the GammaPoisson. However, it appears to be numerically unstable for large counts / small rates. (As a side note, it could be nice to expose the negative binomial as a named distribution.)

Here is an example:
~~~python
latents = [1e4, 1e5, 1e6, 1e7, 1e8, 1e9, 1e10]
fig, ax = plt.subplots(nrows=len(latents), ncols=3, figsize=(10,10))

conc = 500.
p = 0.3

for latent, a in zip(latents, ax):
        
    d1 = dist.GammaPoisson(conc, conc / (latent * p))
    d2 = dist.Poisson(latent * p)
    d3 = dist.Binomial(latent, p)
        
    x = np.linspace(0, latent, 1000)

    a[0].plot(x, d1.log_prob(x))
    a[1].plot(x, d2.log_prob(x))
    a[2].plot(x, d2.log_prob(x))

labels = ['NegBin', 'Pois', 'Binomial']
for label, a in zip(labels, ax[0]):
    a.set_title(label)
    
plt.show()
~~~

This produces:

![image](https://user-images.githubusercontent.com/5331486/78684646-b7bda580-78be-11ea-97c1-197db13d0290.png)

**Comment by fehiepsi** [comment]
@dsheldon Thanks for a clear explanation! I think the issue is JAX's `gammaln` loses precision for high values in 32-bit mode.
```python
import scipy.special as sp
from jax.scipy.special import betaln
x, y = 1e3, 1e9
betaln(x, y), sp.betaln(x, y)
```
I think you can turn on `numpyro.enable_x64()` to improve the precision.

**Comment by dsheldon** [comment]
Hi @fehiepsi, this makes sense, and using `numpyro.enable_x64()` does resolve the issue. I was surprised not to see this behavior with the other distributions (e.g. Binomial) that would also use these functions, but they do indeed appear under the right circumstances, e.g.:

~~~python
n = 1e10
p = 0.0001
x = np.linspace(0, n*p, 1000)
d = dist.Binomial(n, p)
plt.plot(x, d.log_prob(x))
~~~

**Comment by fehiepsi** [comment]
@dsheldon If you observe something strange, feel free to open an issue like this. Thanks for dive into those numerical issues!
